### PR TITLE
[SDESK-4075] Manage Locations

### PR DESCRIPTION
**Pull request by marwoodandrew**

### CODE REVIEW

File: client/controllers/LocationsController.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+    loadWorkspace(store, workspaceChanged) {
+        this.store = store;
+        registerNotifications(this.$scope, this.store);
+
+        if (!workspaceChanged) {
+            return Promise.resolve();
+        }
+
+        // Todo dispatch here if needed;
```
Comment (MarkLark86) [inline client/controllers/LocationsController.js]: ?

File: client/actions/locations.js [hunk 2]
```diff
@@ -1,5 +1,112 @@
...
+import {formatAddress, getErrorMessage} from '../utils';
 import {get} from 'lodash';
+import {MODALS} from '../constants';
+import {showModal} from './index';
+
+
+const locationsSearch = (searchQuery) => (dispatch) => {
+    dispatch(
+        {
+            type: 'LOCATIONS_CHANGE_LIST_SETTINGS',
```
Comment (MarkLark86) [inline client/actions/locations.js]: Create a constants file for location actions (and use these constants in both actions and reducers).

File: client/actions/locations.js [hunk 3]
```diff
@@ -1,5 +1,112 @@
...
+                },
+                (error) => {
+                    notify.error(getErrorMessage(error, gettext('Failed to updated location!')));
+                    return Promise.reject(error);
+                });
+        } else {
+            return api('locations').save({}, updated)
+                .then((newloctaion) => {
+                    dispatch(self.clearEdits());
+                    dispatch(self.searchLocations(getState().locations.searchQuery));
```
Comment (MarkLark86) [inline client/actions/locations.js]: Use `getLocationSearchQuery` selector

File: client/actions/locations.js [hunk 4]
```diff
@@ -1,5 +1,112 @@
...
+    }
+);
+
+const getMoreLocations = () => (
+    (dispatch, getState) => {
+        dispatch({
+            type: 'START_LOADING_LOCATIONS',
+            payload: '',
+        }),
+        dispatch(self.getLocation(getState().locations.searchQuery, false, getState().locations.pageNum + 1))
```
Comment (MarkLark86) [inline client/actions/locations.js]: Use `getLocationSearchQuery` selector.

Also, create a selector that will return the page number, and use that selector here. (Always use selectors when getting data from the store. If the store schema changes you only change the selector instead)

File: client/actions/locations.js [hunk 5]
```diff
@@ -1,5 +1,112 @@
-import {formatAddress} from '../utils';
+import {formatAddress, getErrorMessage} from '../utils';
```
[the remark this answers is not in the document]
Comment (marwoodandrew) [reply client/actions/locations.js]: Done!

File: client/actions/locations.js [hunk 6]
```diff
@@ -1,5 +1,112 @@
...
+    }
+);
+
+
+const deleteLocationConfirmation = (location) => (
+    (dispatch, getState, {api}) => {
+        dispatch(showModal({
+            modalType: MODALS.CONFIRMATION,
+            modalProps: {
+                body: gettext(`Do you want to delete the location \"${location.name}\"?`),
```
[the remark this answers is not in the document]
Comment (marwoodandrew) [reply client/actions/locations.js]: Done!

File: client/actions/locations.js [hunk 7]
```diff
@@ -1,5 +1,112 @@
-import {formatAddress} from '../utils';
+import {formatAddress, getErrorMessage} from '../utils';
 import {get} from 'lodash';
+import {MODALS} from '../constants';
+import {showModal} from './index';
+
+
+const locationsSearch = (searchQuery) => (dispatch) => {
```
[the remark this answers is not in the document]
Comment (marwoodandrew) [reply client/actions/locations.js]: Done!

File: client/components/Locations/LocationsEditor.jsx [hunk 8]
```diff
@@ -0,0 +1,296 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import {SidePanel, Header, Content} from '../UI/SidePanel';
+import {connect} from 'react-redux';
+import * as selectors from '../../selectors';
+import {Field, TextInput, SelectInput, SelectInputWithFreeText} from '../UI/Form';
+import {omit, get, isObject} from 'lodash';
+import {Footer} from '../UI/Popup';
+import {ButtonList} from '../UI';
+import {formatAddress} from '../../utils';
```
[the remark this answers is not in the document]
Comment (marwoodandrew) [reply client/components/Locations/LocationsEditor.jsx]: Done!

File: client/components/Locations/LocationsList.jsx [hunk 9]
```diff
@@ -0,0 +1,92 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import {connect} from 'react-redux';
+import * as selectors from '../../selectors';
+import * as actions from '../../actions';
+import {List} from '../UI';
+import {formatAddress, getMapUrl} from '../../utils';
```
[the remark this answers is not in the document]
Comment (marwoodandrew) [reply client/components/Locations/LocationsList.jsx]: Done!

File: client/components/Locations/LocationsSubNav.jsx [hunk 10]
```diff
@@ -0,0 +1,64 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import {SubNav} from '../UI/SubNav';
+import {SearchBox, Button} from '../UI';
+import {connect} from 'react-redux';
+import * as selectors from '../../selectors';
+import * as actions from '../../actions';
+
```
[the remark this answers is not in the document]
Comment (marwoodandrew) [reply client/components/Locations/LocationsSubNav.jsx]: Done!

File: client/controllers/LocationsController.js (hunk 1 above)
Comment (marwoodandrew) [reply client/controllers/LocationsController.js]: Remove comment

File: client/actions/locations.js (hunk 2 above)
Comment (marwoodandrew) [reply client/actions/locations.js]: Done!

File: client/actions/locations.js (hunk 3 above)
Comment (marwoodandrew) [reply client/actions/locations.js]: Done!

File: client/actions/locations.js (hunk 4 above)
Comment (marwoodandrew) [reply client/actions/locations.js]: Done!

**Pull request merged**